### ISSUE TITLE
AttributeError: 'list' object has no attribute 'infer'

### ISSUE THREAD

**Issue by evandrocoan** [davidhalter/jedi#1034]
<details>
<p>

```
2018-01-26 19:45:40,078 DEBUG pyls.config.config:processmessage:137   pyls_hover [hook]
      config: <pyls.config.config.Config object at 0x0000021DB2112F98>
      workspace: <pyls.workspace.Workspace object at 0x0000021DB2112CF8>
      document: <pyls.workspace.Document object at 0x0000021DB24802B0>
      position: {'line': 122, 'character': 39}

2018-01-26 19:45:40,111 ERROR jsonrpc.manager:_get_responses:122 API Exception: {'type': 'AttributeError', 'args': ("'list' object has no attribute 'infer'",), 'message': "'list' object has no attribute 'infer'"}
Traceback (most recent call last):
  File "F:\Python\lib\site-packages\jsonrpc\manager.py", line 112, in _get_responses
    result = method(*request.args, **request.kwargs)
  File "D:\User\Dropbox\SoftwareVersioning\python-language-server\pyls\dispatcher.py", line 19, in wrapped
    return func(*args, **kwargs)
  File "D:\User\Dropbox\SoftwareVersioning\python-language-server\pyls\python_ls.py", line 164, in m_text_document__hover
    return self.hover(textDocument['uri'], position)
  File "D:\User\Dropbox\SoftwareVersioning\python-language-server\pyls\python_ls.py", line 111, in hover
    return self._hook('pyls_hover', doc_uri, position=position) or {'contents': ''}
  File "D:\User\Dropbox\SoftwareVersioning\python-language-server\pyls\python_ls.py", line 45, in _hook
    return hook(config=self.config, workspace=self.workspace, document=doc, **kwargs)
  File "F:\Python\lib\site-packages\pluggy\__init__.py", line 617, in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
  File "F:\Python\lib\site-packages\pluggy\__init__.py", line 222, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "F:\Python\lib\site-packages\pluggy\__init__.py", line 182, in __call__
    return outcome.get_result()
  File "F:\Python\lib\site-packages\pluggy\callers.py", line 76, in get_result
    raise ex[1].with_traceback(ex[2])
  File "F:\Python\lib\site-packages\pluggy\callers.py", line 48, in from_call
    result = func()
  File "F:\Python\lib\site-packages\pluggy\__init__.py", line 180, in <lambda>
    outcome = _Result.from_call(lambda: self.oldcall(hook, hook_impls, kwargs))
  File "F:\Python\lib\site-packages\pluggy\__init__.py", line 216, in <lambda>
    firstresult=hook.spec_opts.get('firstresult'),
  File "F:\Python\lib\site-packages\pluggy\callers.py", line 201, in _multicall
    return outcome.get_result()
  File "F:\Python\lib\site-packages\pluggy\callers.py", line 76, in get_result
    raise ex[1].with_traceback(ex[2])
  File "F:\Python\lib\site-packages\pluggy\callers.py", line 180, in _multicall
    res = hook_impl.function(*args)
  File "D:\User\Dropbox\SoftwareVersioning\python-language-server\pyls\plugins\hover.py", line 10, in pyls_hover
    definitions = document.jedi_script(position).goto_definitions()
  File "F:\Python\lib\site-packages\jedi\api\__init__.py", line 182, in goto_definitions
    definitions = helpers.evaluate_goto_definition(self._evaluator, context, leaf)
  File "F:\Python\lib\site-packages\jedi\api\helpers.py", line 202, in evaluate_goto_definition
    return evaluator.goto_definitions(context, leaf)
  File "F:\Python\lib\site-packages\jedi\evaluate\__init__.py", line 219, in goto_definitions
    return helpers.evaluate_call_of_leaf(context, name)
  File "F:\Python\lib\site-packages\jedi\evaluate\helpers.py", line 74, in evaluate_call_of_leaf
    return context.eval_node(leaf)
  File "F:\Python\lib\site-packages\jedi\evaluate\base_context.py", line 123, in eval_node
    return self.evaluator.eval_element(self, node)
  File "F:\Python\lib\site-packages\jedi\evaluate\__init__.py", line 179, in eval_element
    return self._eval_element_if_evaluated(context, element)
  File "F:\Python\lib\site-packages\jedi\evaluate\__init__.py", line 191, in _eval_element_if_evaluated
    return self._eval_element_cached(context, element)
  File "F:\Python\lib\site-packages\jedi\evaluate\cache.py", line 40, in wrapper
    rv = function(obj, *args, **kwargs)
  File "F:\Python\lib\site-packages\jedi\evaluate\__init__.py", line 195, in _eval_element_cached
    return eval_node(context, element)
  File "F:\Python\lib\site-packages\jedi\debug.py", line 79, in wrapper
    return func(*args, **kwargs)
  File "F:\Python\lib\site-packages\jedi\evaluate\syntax_tree.py", line 46, in wrapper
    return func(context, *args, **kwargs)
  File "F:\Python\lib\site-packages\jedi\evaluate\syntax_tree.py", line 58, in eval_node
    return eval_atom(context, element)
  File "F:\Python\lib\site-packages\jedi\evaluate\syntax_tree.py", line 172, in eval_atom
    search_global=True
  File "F:\Python\lib\site-packages\jedi\_compatibility.py", line 204, in <lambda>
    return lambda *args, **kwargs: self.func(obj, *args, **kwargs)
  File "F:\Python\lib\site-packages\jedi\evaluate\base_context.py", line 140, in py__getattribute__
    return f.find(filters, attribute_lookup=not search_global)
  File "F:\Python\lib\site-packages\jedi\debug.py", line 79, in wrapper
    return func(*args, **kwargs)
  File "F:\Python\lib\site-packages\jedi\evaluate\finder.py", line 64, in find
    types = self._names_to_types(names, attribute_lookup)
  File "F:\Python\lib\site-packages\jedi\evaluate\finder.py", line 157, in _names_to_types
    contexts = ContextSet.from_sets(name.infer() for name in names)
  File "F:\Python\lib\site-packages\jedi\common\context.py", line 34, in from_sets
    sets = list(sets)
  File "F:\Python\lib\site-packages\jedi\evaluate\finder.py", line 157, in <genexpr>
    contexts = ContextSet.from_sets(name.infer() for name in names)
  File "F:\Python\lib\site-packages\jedi\evaluate\filters.py", line 100, in infer
    return tree_name_to_contexts(self.parent_context.evaluator, self.parent_context, self.tree_name)
  File "F:\Python\lib\site-packages\jedi\evaluate\syntax_tree.py", line 479, in tree_name_to_contexts
    types = _remove_statements(evaluator, context, node, tree_name)
  File "F:\Python\lib\site-packages\jedi\evaluate\syntax_tree.py", line 441, in _remove_statements
    return eval_expr_stmt(context, stmt, seek_name=name)
  File "F:\Python\lib\site-packages\jedi\evaluate\syntax_tree.py", line 46, in wrapper
    return func(context, *args, **kwargs)
  File "F:\Python\lib\site-packages\jedi\evaluate\syntax_tree.py", line 225, in eval_expr_stmt
    return _eval_expr_stmt(context, stmt, seek_name)
  File "F:\Python\lib\site-packages\jedi\debug.py", line 79, in wrapper
    return func(*args, **kwargs)
  File "F:\Python\lib\site-packages\jedi\evaluate\syntax_tree.py", line 241, in _eval_expr_stmt
    context_set = context.eval_node(rhs)
  File "F:\Python\lib\site-packages\jedi\evaluate\base_context.py", line 123, in eval_node
    return self.evaluator.eval_element(self, node)
  File "F:\Python\lib\site-packages\jedi\evaluate\__init__.py", line 174, in eval_element
    return self._eval_element_if_evaluated(context, element)
  File "F:\Python\lib\site-packages\jedi\evaluate\__init__.py", line 191, in _eval_element_if_evaluated
    return self._eval_element_cached(context, element)
  File "F:\Python\lib\site-packages\jedi\evaluate\cache.py", line 40, in wrapper
    rv = function(obj, *args, **kwargs)
  File "F:\Python\lib\site-packages\jedi\evaluate\__init__.py", line 195, in _eval_element_cached
    return eval_node(context, element)
  File "F:\Python\lib\site-packages\jedi\debug.py", line 79, in wrapper
    return func(*args, **kwargs)
  File "F:\Python\lib\site-packages\jedi\evaluate\syntax_tree.py", line 46, in wrapper
    return func(context, *args, **kwargs)
  File "F:\Python\lib\site-packages\jedi\evaluate\syntax_tree.py", line 84, in eval_node
    context_set = eval_trailer(context, context_set, trailer)
  File "F:\Python\lib\site-packages\jedi\evaluate\syntax_tree.py", line 141, in eval_trailer
    ContextualizedNode(context, trailer)
  File "F:\Python\lib\site-packages\jedi\common\context.py", line 65, in mapper
    for context in self._set
  File "F:\Python\lib\site-packages\jedi\common\context.py", line 34, in from_sets
    sets = list(sets)
  File "F:\Python\lib\site-packages\jedi\common\context.py", line 65, in <genexpr>
    for context in self._set
  File "F:\Python\lib\site-packages\jedi\evaluate\base_context.py", line 114, in get_item
    result |= getitem(index)
  File "F:\Python\lib\site-packages\jedi\evaluate\context\iterable.py", line 426, in py__getitem__
    return self._lazy_context_list[index].infer()
AttributeError: 'list' object has no attribute 'infer'
```
</p>
</details><br>

**Comment by davidhalter** [comment]
Can you please add the sources? Where did that happen? If it's proprietary code can you give me a hint what the code looks like or what might cause this error?

Maybe you can also print `self._lazy_context_list` on line 426.

**Comment by evandrocoan** [comment]
This happened when I hovering something `pyls_hover [hook]`. But the document name was hidden `document: <pyls.workspace.Document object at 0x0000021DB24802B0>` because I had temporarily replaced `__str__()` by `__repr__()` to see the difference. Now I restored back `__str__()`, then it should log the document path correctly next time.

```
>pip show jedi
Name: jedi
Version: 0.11.1
Summary: An autocompletion tool for Python that can be used for text editors.
Home-page: https://github.com/davidhalter/jedi
Author: David Halter
Author-email: davidhalter88@gmail.com
License: MIT
Location: f:\python\lib\site-packages
Requires: parso
```

Now what source was I do not remember. I only know it was on `position: {'line': 122, 'character': 39}`. I will have to wait until the next time I stumble with this `position: {'line': 122, 'character': 39}`.

**Comment by davidhalter** [comment]
I have reviewed the code for like an hour. Everything seems fine.

I'm closing, not because this isn't a bug, but because this bug report doesn't help with anything. I still believe that there's a bug (maybe it has been fixed already, who knows, there have been quite some changes), but leaving this issue open doesn't help anybody.

So if anybody has something reproducible, I would be very interested.
